enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,494 @@
+
+###### MachineSet resource
+```
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineSet
+metadata:
+  name: <machine_set_name> 
+  namespace: openshift-cluster-api
+spec:
+  clusterName: <cluster_name> 
+  replicas: 1
+  selector:
+    matchLabels:
+      test: test
+  template:
+    metadata:
+      labels:
+        test: test
+    spec:
+      bootstrap:
+         dataSecretName: worker-user-data 
+      clusterName: <cluster_name> 
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: GCPMachineTemplate 
+        name: <machine_set_name> 
+      failureDomain: <failure_domain> 
+```
+###### GCPMachineTemplate
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: GCPMachineTemplate 
+metadata:
+  name: <template_name> 
+  namespace: openshift-cluster-api
+spec:
+  template:
+    spec: 
+      rootDeviceType: pd-ssd
+      rootDeviceSize: 128
+      instanceType: n1-standard-4
+      image: projects/rhcos-cloud/global/images/rhcos-411-85-202203181601-0-gcp-x86-64
+      subnet: <cluster_name>-worker-subnet
+      serviceAccounts:
+        email: <service_account_email_address>
+        scopes:
+          - https://www.googleapis.com/auth/cloud-platform
+      additionalLabels:
+        kubernetes-io-cluster-<cluster_name>: owned
+      additionalNetworkTags:
+        - <cluster_name>-worker
+      ipForwarding: Disabled
+```
+As can be seen, the bootimage becomes part of an `InfrastructureMachineTemplate` object (eg a GCPMachineTemplate), and then the MachineSet references this template and creates new machines from the template. The stub secret is now stored in a `bootstrap` object. Unlike MAPI backed MachineSets, both of them are no longer part of a single `providerSpec` object. 
+
+It is important to note that InfrastructureMachineTemplate is different per platform and is immutable. This will prevent an update in place style approach and would mean that the template would need to be cloned, updated during the clone, and then the MachineSet updated. This is somewhat similar to the approach used in the current MAPI PoC of cloning the `providerSpec` object, updating it and then patching the `MachineSet`. The `bootstrap` object is platform agnostic, making it somewhat simpler to update. 
+
+Based on the observation above, here is a rough outline of what CAPI support would require:
+- CAPI backed MachineSet detection, so the MSBIC knows when to invoke the CAPI path
+- If a boot image update is required, create a new `InfrastructureMachineTemplate` by cloning the existing and updating the boot image reference within. The name of the new `InfrastructureMachineTemplate` object will be generated by hashing the template content. This is consistent with the current CAPI approach to naming new objects.
+- Updating the ignition stub in `bootstrap.dataSecretName` to the managed stub secret(`*-managed`) if needed.
+- CAPI backed MachineSet patching
+
```

Comment:
The flow sounds reasonable to me from a CAPI perspective.
Is there a plan to garbage collect old/unused `InfrastructureMachineTemplate` after the clone/rereference?

Reply:
I hadn't considered this - thanks! Will add it to the list.